packages/module-utils/src/generate-dts-sourcemaps.ts

```diff
@@ -0,0 +1,256 @@
+                column: stateSourcePosition.column,
+            };
+        }
+    }
+
+    return res;
+}
+
+function findTokenForLine(line: number, dstTokens: TokenizedDtsEntry[]) {
+    for (const typedTokens of dstTokens) {
+        for (const token of typedTokens.tokens) {
+            if ('line' in token) {
+                if (token.line === line) {
+                    return { type: typedTokens.type, token: token };
+                }
+            } else if (
+                token.classStates &&
+                token.classStates.length &&
+                token.classStates[0].stateName.line === line
+            ) {
+                return { type: typedTokens.type, token: token };
+            }
+        }
+    }
+
+    return;
+}
+
+function getClassSourceName(targetName: string, classTokens: ClassesToken): string | undefined {
+    for (const classToken of classTokens.tokens) {
+        if (classToken.outputValue?.value === targetName) {
+            return classToken.value;
+        }
+    }
+
+    return;
+}
+
+export function generateDTSSourceMap(dtsContent: string, meta: StylableMeta) {
+    const tokens = tokenizeDTS(dtsContent);
+    const mapping: Record<number, LineMapping> = {};
+    const lines = dtsContent.split('\n');
+    let lastSrcLocation: Position = { line: 0, column: 0 };
+
+    for (const dtsLine of lines.keys()) {
+        const resToken = findTokenForLine(dtsLine, tokens);
+
+        if (resToken) {
+            let currentSrcPosition: Position | undefined;
+            if ('type' in resToken.token) {
+                const tokenName = stripQuotes(resToken.token.value);
+                switch (resToken.type) {
+                    case 'classes':
+                        currentSrcPosition = getClassSrcPosition(tokenName, meta);
+                        break;
+                    case 'vars':
+                        currentSrcPosition = getVarsSrcPosition(tokenName, meta);
+                        break;
+                    case 'stVars':
+                        currentSrcPosition = getStVarsSrcPosition(tokenName, meta);
+                        break;
+                    case 'keyframes':
+                        currentSrcPosition = getKeyframeSrcPosition(tokenName, meta);
+                        break;
+                }
+
+                if (currentSrcPosition) {
+                    mapping[dtsLine] = createLineMapping(
+                        5,
+                        currentSrcPosition.line - lastSrcLocation.line,
+                        currentSrcPosition.column - lastSrcLocation.column
+                    );
+                    lastSrcLocation = { ...currentSrcPosition };
+                }
+            } else if (resToken.type === 'states') {
+                const classSourceName = getClassSourceName(
+                    resToken.token.className.value,
+                    tokens[1] as ClassesToken
+                );
+
+                if (classSourceName) {
```

Comment:
is there a valid source that produce state in a class that cannot be found? maybe raise a warning?

- same question for other symbols with source position that cannot be located

Reply:
this is mostly due to typing concerns, can't really think of a legit case